adapters/play/src/test/scala/caliban/AdapterSpec.scala

```diff
@@ -0,0 +1,221 @@
+    String.format("%032x", new BigInteger(1, b))
+}
+
+case class UploadFileArgs(file: Upload)
+case class UploadFilesArgs(files: List[Upload])
+
+object TestAPI extends GenericSchema[Blocking with Uploads with Console with Clock] {
+  val api: GraphQL[Blocking with Uploads with Console with Clock] =
+    graphQL(
+      RootResolver(
+        Queries(args => UIO("stub")),
+        Mutations(args => Service.uploadFile(args.file), args => Service.uploadFiles(args.files))
+      )
+    )
+
+  implicit val uploadFileArgsSchema = gen[UploadFileArgs]
+  implicit val mutationsSchema      = gen[Mutations]
+  implicit val queriesSchema        = gen[Queries]
+
+  case class File(hash: String, path: String, filename: String, mimetype: String)
+
+  case class Queries(stub: Unit => UIO[String])
+
+  case class Mutations(
+    uploadFile: UploadFileArgs => ZIO[Blocking with Uploads, Throwable, File],
+    uploadFiles: UploadFilesArgs => ZIO[Blocking with Uploads, Throwable, List[File]]
+  )
+}
+
+object AdapterSpec extends DefaultRunnableSpec {
+  val httpCliLayer: ZLayer[Any, Throwable, HttpClient] =
+    HttpClient.makeHttpClient.toLayer >>> HttpClient.make("127.0.0.1", 8088)
+  val interpLayer: ZLayer[Any, Any, Console with Clock] = Console.live ++ Clock.live
+  val runtime: Runtime[Console with Clock with Blocking with Random with Uploads] =
+    Runtime.unsafeFromLayer(
+      Console.live ++ Clock.live ++ Blocking.live ++ Random.live ++ Uploads.Service.empty,
+      Platform.default
+    )
+
+  // THIS BLOCKS INDEFINTELY DUE TO httpCliLayer
+  //  val runtime: Runtime[HService with HttpClient with Console with Clock] =
+  //    Runtime.unsafeFromLayer(
+  //      (Service.make ++ httpCliLayer ++ Console.live ++ Clock.live),
+  //      Platform.default
+  //    )
+
+  val interpreter: GraphQLInterpreter[Blocking with Uploads with Console with Clock, CalibanError] =
+    runtime.unsafeRun(TestAPI.api.interpreter)
+  val server = AkkaHttpServer.fromRouterWithComponents(
+    ServerConfig(
+      mode = Mode.Dev,
+      port = Some(8088),
+      address = "127.0.0.1"
+    )
+  ) { components =>
+    PlayRouter(
+      interpreter,
+      DefaultControllerComponents(
+        components.defaultActionBuilder,
+        components.playBodyParsers,
+        components.messagesApi,
+        components.langs,
+        components.fileMimeTypes,
+        components.executionContext
+      )
+    )(runtime, components.materializer).routes
+  }
+
+  override def spec: ZSpec[TestEnvironment, Any] =
+    suite("Requests")(
+      testM("multipart request with one file") {
+        val fileHash         = "64498927ff9cd735daefebe7175ed1567650399e58648a6b8340f636243962c0"
+        val fileName: String = s"$fileHash.png"
+        val fileURL: URL     = getClass.getResource(s"/$fileName")
+
+        val query: String =
+          """{ "query": "mutation ($file: Upload!) { uploadFile(file: $file) { hash, path, filename, mimetype } }",   "variables": {  "file": "File" }}"""
```

Comment:
Nit: The spec indicates that the usages of the value should be replaced with `null`, thus it should be `"file": null` I believe